design/storage-class-device-set.md

```diff
@@ -0,0 +1,493 @@
+drives.
+
+DeepSea has a documentation of [some example drive
+groups](https://github.com/SUSE/DeepSea/wiki/Drive-Groups#example-drive-group-files).
+A complete specification is documented in the [ceph
+documentation](http://docs.ceph.com/docs/master/mgr/orchestrator_modules/#orchestrator.DriveGroupSpec).
+
+
+#### DeviceSet vs DriveGroup
+
+A DeviceSet to provision 8 OSDs on 8 drives could look like:
+
+```yaml
+name: "my_device_set"
+count: 8
+```
+
+The Drive Group would look like so:
+
+```yaml
+host_pattern: "hostname1"
+data_devices:
+  count: 8
+```
+
+A Drive Group with 8 OSDs using a shared fast drive could look similar to this:
+
+```yaml
+host_pattern: "hostname1"
+data_devices:
+  count: 8
+  model: MC-55-44-XZ
+db_devices:
+  model: NVME
+db_slots: 8
+```
+#### ResourceRequirements and Placement
+
+Drive Groups don't yet provide orchestrator specific extensions, like resource
+requirements or placement specs, but that could be added trivially. Also a name
+could be added to Drive Groups.
+
+### OSD Configuration Examples
+
+Given the complexity of this design, here are a few examples to showcase
+possible configurations for OSD `StorageClassDeviceSets`.
+
+#### Example 1: AWS cross-AZ
+
+```yaml
+type: CephCluster
+name: cluster1
+...
+spec:
+  ...
+  storage:
+    ...
+    storageClassDeviceSet:
+    - name: cluster1-set1
+      count: 3
+      resources:
+        requests:
+          cpu: 2
+          memory: 4Gi
+      placement:
+        podAntiAffinity:
+          preferredDuringSchedulingIgnoredDuringExecution:
+          - weight: 100
+            podAffinityTerm:
+              labelSelector:
+                matchExpressions:
+                - key: "rook.io/cluster"
+                  operator: In
+                  values:
+                    - cluster1
+                topologyKey: "failure-domain.beta.kubernetes.io/zone"
+      volumeClaimTemplates:
+      - resources:
+          requests:
+            storage: 5Ti
+        storageClassName: gp2-ebs
+```
+
+In this example, `podAntiAffinity` is used to spread the OSD Pods across as many
+AZs as possible. In addition, all Pods would have to be given the label
+`rook.io/cluster=cluster1` to denote they belong to this cluster, such that the
+scheduler will know to try and not schedule multiple Pods with that label on the
+same nodes if possible. The CPU and memory requests would allow the scheduler to
+know if a given node can support running an OSD process.
+
+It should be noted, in the case where the only nodes that can run a new OSD Pod
+is are nodes with OSD Pods already on them, one of those nodes would be
+selected. In addition, EBS volumes may not cross between AZs once greated, so a
+given Pod is guaranteed to always be limited to the AZ it was created in.
+
+#### Example 2: Single AZ
+
+```yaml
+type: CephCluster
+name: cluster1
+...
+spec:
+  ...
+  storage:
+    ...
+    storageClassDeviceSet:
+    - name: cluster1-set1
+      count: 3
+      resources:
```

Comment:
In this example all the cpu and memory resources are the same across the device sets, so how about if you use the single resource definition at the root level to show how the override would work?